### PR TITLE
#6007: Update to RegExp lastIndex description

### PR DESCRIPTION
**Pull request by Prerna237**
> What was wrong/why is this fix needed? (quick summary only)
 A case was missing in the description. 

> Issue number (if there is an associated issue)
https://github.com/mdn/content/issues/6007

> Anything else that could help us review it

### CODE REVIEW

**Comment by wbamberg** [comment]
Is it just me or is this a confusing list?

These kinds of lists of conditions benefit from really precise and consistent language and a lot of clarity around

(1) what the initial conditions are (i.e. value of `lastIndex` and length of input),
(2) what the action is (i.e. calling `exec()`) and
(3) what the outcome is (i.e. a match and/or an updated value of `lastIndex`). 

This list doesn't really achieve that.

It uses different language for what (might be) the same thing "test() and exec() fail" and "the regular expression mismatches input". Some bullets say how `lastIndex` will get updated, and some don't. Some bullets say whether a match is going to be found, and some don't. I'm also not really sure special-casing the empty string is helpful here, isn't that just implicit in the matching rules?

What if we said something like this:

******
When `exec()` or `test()` are called on a given input:

 * If `lastIndex` is greater than the length of the input, `exec()` or `test()` will not find a match, and `lastIndex` will be set to 0.
 * If `lastIndex` is equal to or less than the length of the input, `exec()` or `test()` will attempt to match the input starting from `lastIndex`.
    * If `exec()` or `test()` find a match, then `lastIndex` will be set to the position in the input of the end of the matched string.
    * If `exec()` or `test()` do not find a match, then `lastIndex` will be set to 0.
******

Does that cover all the cases, or am I missing something? This does only implicitly cover the case where `lastIndex` is equal to the input length. So we could add something like:

******
Note that if `lastIndex` is equal to the length of the input, then `exec()` or `test()` will only find a match if they match the empty string.
*******

...but I'm not sure it is necessary, as that's implicit in the regex matching rules.

**Comment by Prerna237** [comment]
I agree. The first time I saw this list I too had a tough time wrapping my head around it. 
Moreover, the bit about matching empty string is redundant. The cases you've mentioned seems to cover it all. Checked. 

I have updated the PR with a rewrite of the rules.

File: files/en-us/web/javascript/reference/global_objects/regexp/lastindex/index.html [hunk 1]
```diff
@@ -21,13 +21,13 @@
...
- <li>If <code>lastIndex</code> is equal to or less than the length of the string and if the regular expression matches the empty string, then the regular expression matches input starting from <code>lastIndex</code>.</li>
- <li>If <code>lastIndex</code> is equal to the length of the string and if the regular expression does not match the empty string, then the regular expression mismatches input, and <code>lastIndex</code> is reset to 0.</li>
- <li>Otherwise, <code>lastIndex</code> is set to the next position following the most recent match.</li>
+<li>If <code>lastIndex</code> is greater than the length of the input, exec() or test() will not find a match, and <code>lastIndex</code> will be set to 0.</li>
+<li>If <code>lastIndex</code> is equal to or less than the length of the input, exec() or test() will attempt to match the input starting from <code>lastIndex</code>.</li>
+<li>If exec() or test() find a match, then <code>lastIndex</code> will be set to the position of the end of the matched string in the input.</li>
+<li>If exec() or test() do not find a match, then <code>lastIndex</code> will be set to 0.</li>
```
Comment (wbamberg) [inline files/en-us/web/javascript/reference/global_objects/regexp/lastindex/index.html:30]:
```suggestion
<li>If <code>lastIndex</code> is greater than the length of the input, <code>exec()</code> or <code>test()</code> will not find a match, and <code>lastIndex</code> will be set to 0.</li>
<li>If <code>lastIndex</code> is equal to or less than the length of the input, <code>exec()</code> or <code>test()</code> will attempt to match the input starting from <code>lastIndex</code>.
  <ul>
    <li>If <code>exec()</code> or <code>test()</code> find a match, then <code>lastIndex</code> will be set to the position of the end of the matched string in the input.</li>
    <li>If <code>exec()</code> or <code>test()</code> do not find a match, then <code>lastIndex</code> will be set to 0.</li>
  </ul>
</li>
```

**Review by wbamberg — APPROVED** [review]
Sorry to be slow reviewing, @Prerna237 , I was on vacation. Thanks a lot for making the updates! I tweaked it a bit to add code formatting and make the two results of attempting a match be a sub-list.

**Pull request merged**